### PR TITLE
T20: Various Tablet improvements

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #1597

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 1]
```diff
@@ -165,49 +162,52 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
-        threadListObserver?.let(threadListAdapter::registerAdapterDataObserver)
     }
 
     private fun navigateFromNotificationToNewMessage() {
-        // Here, we use `arguments` instead of `navigationArgs` because we need mutable data.
-        if (arguments?.getString(navigationArgs::replyToMessageUid.name) != null) {
-            // If we are coming from the Reply action of a Notification, we need to navigate to NewMessageActivity.
+        arguments?.consumeKeyIfProvided(navigationArgs::replyToMessageUid.name) { replyToMessageUid ->
+
+            // If we clicked on the "Reply" action of a Notification, we need to navigate to ThreadFragment.
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:187]: Suspicious

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 2]
```diff
@@ -48,9 +53,12 @@ abstract class TwoPaneFragment : Fragment() {
             else -> throw IllegalStateException()
         }
 
+    open fun doAfterFolderChanged() {}
+
+    fun isOnlyOneShown() = slidingPaneLayout.isSlideable
     fun areBothShown() = !slidingPaneLayout.isSlideable
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:59]: not a fan of this 2 functions being a duplicate

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 3]
```diff
@@ -63,46 +71,77 @@ abstract class TwoPaneFragment : Fragment() {
...
 
-    private fun observeThreadEvents() = with(mainViewModel) {
+    private fun observeThreadUid() {
 
-        val threadListAdapter = when (this@TwoPaneFragment) {
-            is ThreadListFragment -> this@TwoPaneFragment.threadListAdapter
-            is SearchFragment -> this@TwoPaneFragment.threadListAdapter
+        val threadListAdapter = when (this) {
+            is ThreadListFragment -> this.threadListAdapter
+            is SearchFragment -> this.threadListAdapter
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:129]:
```suggestion
        val threadListAdapter = when (this) {
            is TwoPaneFragment -> this.threadListAdapter
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 4]
```diff
@@ -63,46 +71,77 @@ abstract class TwoPaneFragment : Fragment() {
...
         }
     }
 
-    private fun observeThreadEvents() = with(mainViewModel) {
+    private fun observeThreadUid() {
 
-        val threadListAdapter = when (this@TwoPaneFragment) {
-            is ThreadListFragment -> this@TwoPaneFragment.threadListAdapter
-            is SearchFragment -> this@TwoPaneFragment.threadListAdapter
+        val threadListAdapter = when (this) {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:127]: It is only used in the else, put this there

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        newMessageArgs.value = NewMessageActivityArgs(
+            draftMode = draftMode,
+            previousMessageUid = previousMessageUid,
+            shouldLoadDistantResources = shouldLoadDistantResources,
+            arrivedFromExistingDraft = arrivedFromExistingDraft,
+            draftLocalUuid = draftLocalUuid,
+            draftResource = draftResource,
+            messageUid = messageUid,
+            mailToUri = mailToUri,
+        )
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt:107]: parameters' names aren't really useful here

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt [hunk 6]
```diff
@@ -26,11 +26,13 @@ import com.infomaniak.lib.core.utils.safeBinding
 import com.infomaniak.mail.R
 import com.infomaniak.mail.databinding.BottomSheetActionsMenuBinding
 import com.infomaniak.mail.ui.MainViewModel
+import com.infomaniak.mail.ui.main.folder.TwoPaneViewModel
 
 abstract class MailActionsBottomSheetDialog : ActionsBottomSheetDialog() {
 
     protected var binding: BottomSheetActionsMenuBinding by safeBinding()
     protected val mainViewModel: MainViewModel by activityViewModels()
+    protected val twoPaneViewModel: TwoPaneViewModel by activityViewModels()
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt:35]: Not used anymore

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 7]
```diff
@@ -48,9 +53,12 @@ abstract class TwoPaneFragment : Fragment() {
             else -> throw IllegalStateException()
         }
 
+    open fun doAfterFolderChanged() {}
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:56]:
```suggestion
    open fun doAfterFolderChanged() = Unit
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 2 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:61]: Do you prefer this ?
```kt
fun isOnlyOneShown() = slidingPaneLayout.isSlideable
fun areBothShown() = !isOnlyOneShown()
```
🤷‍♂️

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt (hunk 5 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt:106]: Try to remove them, then come back to me :D

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt (hunk 6 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt:35]: No?!

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 7 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:58]: Nope, it's only used in one of both children.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 3 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt]: Handled by injection now, much simpler.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt]: Handled by injection now, much simpler.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt (hunk 6 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt:35]: 🦔 🐛

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt (hunk 7 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt]: Then make it return Unit

**Review by FabianDevel — APPROVED** [review]

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt [hunk 8]
```diff
@@ -119,44 +147,39 @@ abstract class TwoPaneFragment : Fragment() {
...
 
-    fun openThread(uid: String) {
-
-        mainViewModel.currentThreadUid.value = uid
-
-        val isOpening = slidingPaneLayout.openPane()
+    fun navigateToThread(thread: Thread) {
+        if (thread.isOnlyOneDraft) {
+            trackNewMessageEvent(OPEN_FROM_DRAFT_NAME)
+            openDraft(thread)
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneFragment.kt:148]: 🤔 Did you read what `openDraft()` is doing ?

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    fun openThread(uid: String) {
+        currentThreadUid.value = uid
+    }
+
+    fun closeThread() {
+        currentThreadUid.value = null
+    }
+
+    fun openDraft(thread: Thread) {
+        navigateToSelectedDraft(thread.messages.first())
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt:74]: Can we use `.single()` instead of `.first()` ?

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt (hunk 9 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/TwoPaneViewModel.kt:74]: Yes we can.

**Pull request merged**